### DISCUSSION TITLE
REST API v3: wildcard branch protection

### DISCUSSION THREAD

**Discussion by beiriannydd** [community/community#24703] [Apps, API and Webhooks]
<p>Is it possible to GET/SET wildcard branch protection using the API?  Querying does not seem to work with wildcards at least, although branches being matched are individually shown as being protected, it doesn’t seem possible to tell the difference between ones matching a wildcard, and those with specific settings.</p>

**Comment by nadiajoyce** [comment, marked as answer]
<p>Hi <a class="mention" href="/beiriannydd">@beiriannydd</a>,</p>
<p>Thanks for being here and for asking this question. At this time, this isn’t supported. However, we’re always working to improve, and we consider every suggestion we receive. I’ve logged this in our internal feature request list. Though I can’t guarantee anything or share a timeline for this, I can tell you that it’s been shared with the appropriate teams for consideration.</p>
<p>Please let me know if you have any other questions.</p>
<p>Cheers!</p>

**Comment by jonathanan** [comment]
<p>Thanks <a class="mention" href="/nadiajoyce">@nadiajoyce</a> . A use case here would be using the terraform github provider. </p>
<p><a href="https://github.com/terraform-providers/terraform-provider-github/issues/164" rel="nofollow noopener">https://github.com/terraform-providers/terraform-provider-github/issues/164</a></p>

**Comment by otter42** [comment]
<p>Unfortunate that the api support didn’t come in advance of or with the ux support… makes it hard to take advantage of when programatically creating the repositories on behalf of someone who wont have rights to create the restrictions.  Hopefully it will be coming in the not too distant future.</p>

**Comment by maheshika** [comment]
<p>This would be a great feature for organizations which have multiple repos and branches.</p>

**Reply by maheshika** [reply]
Since the wildcard is not supported, I had to find a workaround [1]. The disadvantage is that we have to run this script very often and it protects all branches. However, that may be a better way to do this.

https://medium.com/@maheshika/protecting-multiple-branches-ec3156a5e5a2

**Comment by tadms** [comment]
<p><a class="mention" href="/nadiajoyce">@nadiajoyce</a> - Is there any update on potentially enabling this feature?  This is a significant painpoint for migrating my Organization (130+ repos) to GitHub.  As is, I’m stuck configuring this manually for each repo as we have a hard requirement for wildcard branch protection.  It would be great to be able to automate this.  </p>

**Comment by JackFreelander** [comment]
<p>+1 for this being a real nuisance and gap in the octokit APIs.  It makes no sense that these can be created, edited and deleted manually using the UI, but there is no programmatic mechanism for doing so.</p>
<p>Can you please explain the technical details regarding why this is difficult or hasn’t been implemented?</p>

**Comment by DaWyz** [comment]
<p>Hello, I know this question is related to the Rest API v3 and you may have to stick with it but there is a <a href="https://developer.github.com/v4/input_object/createbranchprotectionruleinput/" rel="nofollow noopener">mutation available using the octokit GraphQL API V4</a>. It supports actual pattern such as <strong>release/*</strong> or <strong>hotfix/*</strong>.</p>
<p>See example below:</p>
<pre><code>const { graphql } = require('@octokit/graphql');

const octoql = graphql.defaults({
      headers: {
        authorization: `token ${process.env.GITHUB_TOKEN}`
      }
});

octoql(
          `
            mutation($input: CreateBranchProtectionRuleInput!) {
              createBranchProtectionRule(input: $input) {
                branchProtectionRule {
                  id
                }
              }
            }
          `,
          {
            input: {
              repositoryId: repo.data.node_id,
              pattern: "release/*",
              dismissesStaleReviews: true,
              requiresApprovingReviews: true,
              requiredApprovingReviewCount: 1,
              isAdminEnforced: true,
              requiresStatusChecks: null
            }
          }
        );
</code></pre>
<p>You can get the repositoryId using the Rest API or the GraphQL API.</p>
<p>Hope this helps.</p>

**Comment by tadms** [comment]
<p>For anyone trying to work GraphQL into a cURL call / shell script (like I was), after a lot of pain I eventually found success with something like this.  You’ll have to rework it for the protections you need, and obviously replace the ‘xxxxxxx’ values with your actual hex IDs.  Hopefully this helps someone, I wasn’t able to find examples of this anywhere. </p>
<pre><code>function generate_gh_wildcard_branch_post_data() {
local repo_node_id="xxxxxxxxxxxxxxxxxxxxxxxx"
local pattern="release/*"
local admin_team_node_id="xxxxxxxxxxxxxxxxxxxxxxxxx"
local write_team_node_id="xxxxxxxxxxxxxxxxxxxxxxxxx"
cat &lt;&lt;EOF
{"query": "mutation CreateBranchProtectionRule { createBranchProtectionRule(input:{repositoryId:\"$repo_node_id\" pattern:\"$pattern\" requiresApprovingReviews: true requiredApprovingReviewCount: 2 requiresCodeOwnerReviews: true restrictsPushes: true pushActorIds: [\"$admin_team_node_id\" \"$write_team_node_id\"] dismissesStaleReviews: true}) { clientMutationId } }"}
EOF
}

function github_add_wildcard_branch_protection {
  local payload="$(generate_gh_wildcard_branch_post_data)"
  curl -ks -X POST \
    -H 'Content-Type: application/json' \
    -H "Authorization: bearer $GITHUB_TOKEN" \
    --data "$payload" \
    $GITHUB_GRAPHQL_API
}
</code></pre>

**Comment by llaski** [comment]
<p>Is there any update on this being added to the GitHub Rest API? GitLab solved this issue over 2 years ago and this is a pretty large pain point for our organization in having to update hundreds of repositories manually so they all have the same branch protection rules. For reference we don’t have time/skillset to implement the GraphQL solution and believe this should be supported in the Rest API since it’s possible through the normal web UI.</p>

**Comment by rmoles** [comment]
<p>Is this request on the radar for GitHub? It would be an extremely helpful feature!</p>

**Comment by HenryHoggard** [comment]
<p>Any update on this and whether its been added to the roadmap?</p>

**Comment by Danielku15** [comment]
<p>I would be also interested in an update on this API. It feels really strange that the API looks so different from how you configure things in the UI. Likely a historical thing but still, there is no way to reproduce the “clicks” you do on the UI.</p>
<p>An example: we have one branch protection for <code>release/*</code> requiring review by the release manager on all our release branches. We would like to create/maintain this rule via API to roll it out across all our repos consistently. I see no way to achieve this by today. Any API action, will mess with previously UI created rules instead of just adapting settings.</p>

**Comment by gms8994** [comment]
Is there any chance of an update from Github on this? We need to make a similar branch protection rule across dozens of repositories that are all based on a wildcard branch name.

**Comment by L1ghtman2k** [comment]
I want to be able to list all branch protection rules on the repo. ThIs currently seems not to be possible with the REST API

**Comment by robin-coste** [comment]
I also have the `release/*` workflow, would there be any updates on this issue ? :)

**Comment by patrick-knight** [comment]
Hey y'all,

Probably not the _exact_ answer everyone is looking for but we recently announced the public beta of [Repository rules](https://github.blog/changelog/2023-04-17-introducing-repository-rules-public-beta/), which is intended to be the future of protected branches. 

One of the _subtle_ changes for most but relevant for y'all is we have ensured [the REST endpoint](https://docs.github.com/en/rest/repos/rules?apiVersion=2022-11-28#create-a-repository-ruleset) does support wildcards as well as multiple conditions for matching branches.


```

"conditions": {
    "ref_name": {
      "include": [
        "refs/heads/release*",
        "refs/heads/main"
      ],
      "exclude": [
        "refs/heads/dev*"
      ]
    }
```
